contracts/AelinUpFrontDeal.sol

```diff
@@ -400,41 +410,47 @@ contract AelinUpFrontDeal is AelinERC20, MinimalProxyFactory, IAelinUpFrontDeal
+    function claimUnderlying(uint256 _tokenId) external {
+        _claimUnderlying(msg.sender, _tokenId);
+    }
+
+    function _claimUnderlying(address _owner, uint256 _tokenId) internal purchasingOver passMinimumRaise {
+        require(ownerOf(_tokenId) == _owner, "must be owner to claim");
+        uint256 claimableAmount = claimableUnderlyingTokens(_tokenId);
+        require(claimableAmount > 0, "no underlying ready to claim");
         address _underlyingDealToken = dealData.underlyingDealToken;
-        amountVested[msg.sender] += underlyingDealTokensClaimed;
-        _burn(msg.sender, underlyingDealTokensClaimed);
-        totalUnderlyingClaimed += underlyingDealTokensClaimed;
-        IERC20(_underlyingDealToken).safeTransfer(msg.sender, underlyingDealTokensClaimed);
-        emit ClaimedUnderlyingDealToken(msg.sender, _underlyingDealToken, underlyingDealTokensClaimed);
+        tokenDetails[_tokenId].lastClaimedAt = block.timestamp;
+        amountVested[msg.sender] += claimableAmount;
+        totalUnderlyingClaimed += claimableAmount;
+        IERC20(_underlyingDealToken).safeTransfer(_owner, claimableAmount);
+        emit ClaimedUnderlyingDealToken(_owner, _underlyingDealToken, claimableAmount);
     }
 
     /**
-     * @dev a view showing the amount of the underlying deal token a purchaser can claim
-     * @param _purchaser address to check the quantity of claimable underlying tokens
+     * @dev a view showing the amount of the underlying deal token a ERC721 deal token holder can claim
+     * @param _tokenId the token ID to check the quantity of claimable underlying tokens
      */
-    function claimableUnderlyingTokens(address _purchaser) public view purchasingOver returns (uint256) {
-        uint256 _vestingPeriod = dealConfig.vestingPeriod;
-        uint256 precisionAdjustedUnderlyingClaimable;
 
+    function claimableUnderlyingTokens(uint256 _tokenId) public view returns (uint256) {
+        TokenDetails memory schedule = tokenDetails[_tokenId];
         uint256 maxTime = block.timestamp > vestingExpiry ? vestingExpiry : block.timestamp;
-        if (
-            balanceOf(_purchaser) > 0 &&
-            (maxTime > vestingCliffExpiry || (maxTime == vestingCliffExpiry && _vestingPeriod == 0))
-        ) {
-            uint256 timeElapsed = maxTime - vestingCliffExpiry;
-
-            uint256 underlyingClaimable = _vestingPeriod == 0
-                ? balanceOf(_purchaser)
-                : ((balanceOf(_purchaser) + amountVested[_purchaser]) * timeElapsed) /
-                    _vestingPeriod -
-                    amountVested[_purchaser];
+        uint256 minTime = schedule.lastClaimedAt > vestingCliffExpiry ? schedule.lastClaimedAt : vestingCliffExpiry;
+        uint256 vestingPeriod = dealConfig.vestingPeriod;
+        uint256 precisionAdjustedUnderlyingClaimable;
+
+        if (maxTime > vestingCliffExpiry && minTime <= vestingExpiry) {
```

Comment:
it looks like this new logic is actually good. I will take another look tomorrow. the only problem is if you call it on a token id that doesn't exist. Also, we can probably remove amountVested mapping from the client if we are not tying it to their address and are just tracking the last time they claimed and the amount they have to vest since then since we are not using balanceOf anymore

Reply:
agreed I don't think we need `amountVested` anymore. If we want to track it in the UI we can just use the graph